tests/all/epoch_interruption.rs

```diff
@@ -0,0 +1,124 @@
+use std::future::Future;
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use wasmtime::*;
+
+fn dummy_waker() -> Waker {
+    return unsafe { Waker::from_raw(clone(5 as *const _)) };
+
+    unsafe fn clone(ptr: *const ()) -> RawWaker {
+        assert_eq!(ptr as usize, 5);
+        const VTABLE: RawWakerVTable = RawWakerVTable::new(clone, wake, wake_by_ref, drop);
+        RawWaker::new(ptr, &VTABLE)
+    }
+
+    unsafe fn wake(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+
+    unsafe fn wake_by_ref(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+
+    unsafe fn drop(ptr: *const ()) {
+        assert_eq!(ptr as usize, 5);
+    }
+}
+
+fn build_engine() -> Engine {
+    let mut config = Config::new();
+    config.async_support(true);
+    config.epoch_interruption(true);
+    Engine::new(&config).unwrap()
+}
+
+fn make_env(engine: &Engine) -> Linker<()> {
+    let mut linker = Linker::new(engine);
+    let engine = engine.clone();
+
+    linker
+        .func_new(
+            "",
+            "bump_epoch",
+            FuncType::new(None, None),
+            move |_caller, _params, _results| {
+                engine.increment_epoch();
+                Ok(())
+            },
+        )
+        .unwrap();
+
+    linker
+}
+
+fn run_and_count_yields(wasm: &str, initial: u64, delta: u64) -> usize {
+    let engine = build_engine();
+    let linker = make_env(&engine);
+    let module = Module::new(&engine, wasm).unwrap();
+    let mut store = Store::new(&engine, ());
+    let instance = linker.instantiate(&mut store, &module).unwrap();
+    let f = instance.get_func(&mut store, "run").unwrap();
+
+    store.set_epoch_deadline_refill_delta(delta);
+    store.refill_epoch_deadline(initial);
+
+    let mut future = Box::pin(f.call_async(&mut store, &[], &mut []));
+    let mut yields = 0;
+    loop {
+        match future
+            .as_mut()
+            .poll(&mut Context::from_waker(&dummy_waker()))
+        {
+            Poll::Pending => {
+                yields += 1;
+            }
+            Poll::Ready(result) => {
+                result.unwrap();
+                break;
+            }
+        }
+    }
+
+    yields
+}
+
+#[test]
+fn epoch_yield_at_func_entry() {
+    assert_eq!(
+        1,
+        run_and_count_yields(
+            // Should yield at start of call to func 2.
+            "
+            (module
+                (import \"\" \"bump_epoch\" (func))
+                (func (export \"run\")
+                    call 0  ;; bump epoch
+                    call 2) ;; call func; will notice new epoch and yield
```

Comment:
Could this use `$foo` for names of functions?

Reply:
Done!